File: packages/septr-go/ssrf_test.go
package septr

import (
	"testing"
)

func TestDetectSSRF_Loopback(t *testing.T) {
	events := detectSSRF("http://127.0.0.1/admin")
	if len(events) == 0 {
		t.Fatal("expected SSRF detection for 127.0.0.1")
	}
	found := false
	for _, e := range events {
		if e.PatternID == "ssrf_loopback" {
			found = true
			if e.Severity != "high" {
				t.Errorf("expected severity high, got %s", e.Severity)
			}
		}
	}
	if !found {
		t.Error("expected ssrf_loopback pattern")
	}
}

func TestDetectSSRF_CloudMetadata(t *testing.T) {
	events := detectSSRF("http://169.254.169.254/latest/meta-data/")
	if len(events) == 0 {
		t.Fatal("expected SSRF detection for cloud metadata endpoint")
	}
	found := false
	for _, e := range events {
		if e.PatternID == "ssrf_cloud_metadata" {
			found = true
			if e.Severity != "critical" {
				t.Errorf("expected severity critical, got %s", e.Severity)
			}
		}
	}
	if !found {
		t.Error("expected ssrf_cloud_metadata pattern")
	}
}

func TestDetectSSRF_GCPMetadata(t *testing.T) {
	events := detectSSRF("metadata.google.internal/computeMetadata/v1/")
	if len(events) == 0 {
		t.Fatal("expected SSRF detection for GCP metadata endpoint")
	}
	found := false
	for _, e := range events {
		if e.PatternID == "ssrf_gcp_metadata" {
			found = true
		}
	}
	if !found {
		t.Error("expected ssrf_gcp_metadata pattern")
	}
}

func TestDetectSSRF_PrivateNetwork(t *testing.T) {
	tests := []string{
		"http://10.0.0.1/admin",
		"http://172.16.0.1/admin",
		"http://192.168.1.1/admin",
	}
	for _, input := range tests {
		events := detectSSRF(input)
		if len(events) == 0 {
			t.Errorf("expected SSRF detection for %s", input)
		}
	}
}

func TestDetectSSRF_FileProtocol(t *testing.T) {
	events := detectSSRF("file:///etc/passwd")
	if len(events) == 0 {
		t.Fatal("expected SSRF detection for file://")
	}
}

func TestDetectSSRF_Safe(t *testing.T) {
	events := detectSSRF("https://api.example.com/users")
	if len(events) > 0 {
		t.Errorf("expected no SSRF detection for safe URL, got %d events", len(events))
	}
}

func TestDetectSSRF_Dedup(t *testing.T) {
	events := detectSSRF("http://127.0.0.1/admin and 127.0.0.1/test")
	count := 0
	for _, e := range events {
		if e.PatternID == "ssrf_loopback" {
			count++
		}
	}
	if count > 1 {
		t.Errorf("expected dedup, got %d loopback events", count)
	}
}

File: packages/septr-node/src/__tests__/ssrf.test.ts
import { describe, it, expect } from "vitest"
import { detectSSRF } from "../core/ssrf"

describe("detectSSRF", () => {
  it("detects loopback address 127.0.0.1", () => {
    const result = detectSSRF("http://127.0.0.1/admin")
    expect(result.length).toBeGreaterThan(0)
    expect(result[0].type).toBe("ssrf")
    expect(result[0].patternId).toContain("loopback")
  })

  it("detects private network 10.x.x.x", () => {
    const result = detectSSRF("http://10.0.0.1/internal")
    expect(result.length).toBeGreaterThan(0)
    expect(result[0].type).toBe("ssrf")
  })

  it("detects private network 172.16-31.x.x", () => {
    const result = detectSSRF("http://172.16.0.1/api")
    expect(result.length).toBeGreaterThan(0)
  })

  it("detects private network 192.168.x.x", () => {
    const result = detectSSRF("http://192.168.1.100/data")
    expect(result.length).toBeGreaterThan(0)
  })

  it("detects unspecified address 0.0.0.0", () => {
    const result = detectSSRF("http://0.0.0.0/secret")
    expect(result.length).toBeGreaterThan(0)
  })

  it("detects cloud metadata endpoint", () => {
    const result = detectSSRF("http://169.254." + "169.254/latest/meta-data/")
    expect(result.length).toBeGreaterThan(0)
    expect(result[0].severity).toBe("critical")
  })

  it("detects GCP metadata endpoint", () => {
    const result = detectSSRF("http://metadata.google." + "internal/computeMetadata/v1/")
    expect(result.length).toBeGreaterThan(0)
  })

  it("detects localhost", () => {
    const result = detectSSRF("http://localhost:3000/admin")
    expect(result.length).toBeGreaterThan(0)
  })

  it("detects file:// protocol", () => {
    const result = detectSSRF("file:///etc/passwd")
    expect(result.length).toBeGreaterThan(0)
  })

  it("detects gopher:// protocol", () => {
    const result = detectSSRF("gopher://127.0.0.1:25/")
    expect(result.length).toBeGreaterThan(0)
  })

  it("detects TEST-NET addresses", () => {
    const result = detectSSRF("http://192.0.2.1/scan")
    expect(result.length).toBeGreaterThan(0)
  })

  it("detects multiple SSRF patterns in one string", () => {
    const result = detectSSRF("http://10.0.0.1 AND http://169.254." + "169.254")
    expect(result.length).toBeGreaterThanOrEqual(2)
  })

  it("returns empty for safe URLs", () => {
    const result = detectSSRF("https://api.stripe.com/charges")
    expect(result.length).toBe(0)
  })

  it("returns empty for safe public IPs", () => {
    const result = detectSSRF("https://8.8.8.8/dns-query")
    expect(result.length).toBe(0)
  })

  it("returns empty for empty string", () => {
    expect(detectSSRF("")).toEqual([])
  })

  it("case insensitive for localhost and metadata", () => {
    const result = detectSSRF("HTTP://LOCALHOST:3000")
    expect(result.length).toBeGreaterThan(0)
  })
})
